crates/plugins/src/badger/mod.rs

```diff
@@ -0,0 +1,318 @@
+        }
+    }
+
+    fn fire_and_forget_update_impl() -> anyhow::Result<()> {
+        let mut update_cmd = tokio::process::Command::new(std::env::current_exe()?);
+        update_cmd.args(["plugins", "update"]);
+        update_cmd.stdout(std::process::Stdio::null());
+        update_cmd.stderr(std::process::Stdio::null());
+        update_cmd.spawn()?;
+        Ok(())
+    }
+
+    async fn check(&self) -> anyhow::Result<BadgerUI> {
+        let available_upgrades = self.available_upgrades().await?;
+
+        // TO CONSIDER: skipping this check and badgering for the same upgrade in case they missed it
+        if self
+            .previous_badger
+            .includes_any(&available_upgrades.list())
+        {
+            return Ok(BadgerUI::None);
+        }
+
+        if !available_upgrades.is_none() {
+            self.record_manager
+                .record_badger(
+                    &self.plugin_name,
+                    &self.current_version,
+                    &available_upgrades.list(),
+                )
+                .await
+        };
+
+        Ok(available_upgrades.classify())
+    }
+
+    async fn available_upgrades(&self) -> anyhow::Result<AvailableUpgrades> {
+        let store = self.plugin_manager.store();
+
+        let latest_version = {
+            let latest_lookup = crate::lookup::PluginLookup::new(&self.plugin_name, None);
+            let latest_manifest = latest_lookup
+                .get_manifest_from_repository(store.get_plugins_directory())
+                .await
+                .ok();
+            latest_manifest.and_then(|m| semver::Version::parse(m.version()).ok())
+        };
+
+        let manifests = store.catalogue_manifests()?;
+        let relevant_manifests = manifests
+            .into_iter()
+            .filter(|m| m.name() == self.plugin_name);
+        let compatible_manifests = relevant_manifests.filter(|m| {
+            m.has_compatible_package() && m.is_compatible_spin_version(self.spin_version)
+        });
+        let compatible_plugin_versions =
+            compatible_manifests.filter_map(|m| PluginVersion::try_from(m, &latest_version));
+        let considerable_manifests = compatible_plugin_versions
+            .filter(|pv| !pv.is_prerelease() && pv.is_higher_than(&self.current_version))
+            .collect::<Vec<_>>();
+
+        let (eligible_manifests, questionable_manifests) = if self.current_version.major == 0 {
+            (vec![], considerable_manifests)
+        } else {
+            considerable_manifests
+                .into_iter()
+                .partition(|pv| pv.version.major == self.current_version.major)
+        };
+
+        let highest_eligible_manifest = eligible_manifests
+            .into_iter()
+            .max_by_key(|pv| pv.version.clone());
+        let highest_questionable_manifest = questionable_manifests
+            .into_iter()
+            .max_by_key(|pv| pv.version.clone());
+
```

Comment:
According to the definition of `Both` being that the questionable is higher than the elgible, do we need a check here of whether `highest_questionable_manifest > highest_eligible_manifest` and if false set `highest_questionable_manifest` to false?

Reply:
`considerable_manifests` contains only manifests whose version is >= current version.  If HEV exists then its major version is == current major version.  If HQV exists, its major version is != current major version.  But since HQV major must be >= current major version, this means HQV major > current major.  And since current major == HEV major, if both exist then HQV major > HEV major.

![image](https://github.com/fermyon/spin/assets/865538/655c85c9-657e-4f63-90c4-d5e825e35078)

I can add a comment explaining this if it would be useful.

Reply:
just cackled 🤣  thank you for laying all this out! I don't think we need to comment it